agent_app_sample_code/01_data_pipeline.py

```diff
@@ -0,0 +1,580 @@
+# MAGIC
+# MAGIC     *Note: While you can adjust the other settings and modify the parsing/chunking code, we suggest doing so only after evaluating your Agent's quality so you can make improvements that specifically address root causes of quality issues.*
+# MAGIC
+# MAGIC 2. **Use later notebooks to intergate the retriever into an the agent and evaluate the agent/retriever's quality.**
+# MAGIC
+# MAGIC 3. **If the evaluation results show retrieval issues as a root cause, use this notebook to iterate on your data pipeline's code & config.** Below are some potential fixes you can try, see the AI Cookbook's [debugging retrieval issues](https://ai-cookbook.io/nbs/5-hands-on-improve-quality-step-1-retrieval.html) section for details.**
+# MAGIC     - Add missing, but relevant source documents into in the index.
+# MAGIC     - Resolve any conflicting information in source documents.
+# MAGIC     - Adjust the data pipeline configuration:
+# MAGIC       - Modify chunk size or overlap.
+# MAGIC       - Experiment with different embedding models.
+# MAGIC     - 3️Adjust the data pipeline code:
+# MAGIC       - Create a custom parser or use different parsing libraries.
+# MAGIC       - Develop a custom chunker or use different chunking techniques.
+# MAGIC       - Extract additional metadata for each document.
+# MAGIC     - Adjust the Agent's code/config in subsequent notebooks:
+# MAGIC       - Change the number of documents retrieved (K).
+# MAGIC       - Try a re-ranker.
+# MAGIC       - Use hybrid search.
+# MAGIC       - Apply extracted metadata as filters.
+# MAGIC
+# MAGIC
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC **Important note:** Throughout this notebook, we indicate which cells you:
+# MAGIC - ✅✏️ *should* customize - these cells contain code & config with business logic that you should edit to meet your requirements & tune quality
+# MAGIC - 🚫✏️ *typically will not* customize - these cells contain boilerplate code required to execute the pipeline
+# MAGIC
+# MAGIC *Cells that don't require customization still need to be run!  You CAN change these cells, but if this is the first time using this notebook, we suggest not doing so.*
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Install Python libraries
+# MAGIC
+# MAGIC 🚫✏️ Only modify if you need additional packages in your code changes to the document parsing or chunking logic.
+
+# COMMAND ----------
+
+# Versions of Databricks code are not locked since Databricks ensures changes are backwards compatible.
+# Versions of open source packages are locked since package authors often make backwards compatible changes
+%pip install -qqqq -U \
+  -r requirements.txt `# Packages shared across all notebooks` \
+  pymupdf4llm==0.0.5 pymupdf==1.24.5 `# PDF parsing` \
+  markdownify==0.12.1  `# HTML parsing` \
+  pypandoc_binary==1.13  `# DOCX parsing` \
+  transformers==4.41.1 torch==2.3.0 tiktoken==0.7.0 langchain-text-splitters==0.2.0. `# For get_recursive_character_text_splitter()`
```

Comment:
customer feedback - it's not uncommon for this to break in customer environments without external internet access. Can we put it behind a flag and toggle the text splitter behavior based on data pipeline config?

Reply:
We could, but we would need to write a default text splitter that doesn't use any external packages.  these are all required by the default splitter :(